File: Práctica/2.5 - RobocodeLabo II/src/CasseraBottaBianchi/EstrategaImplementacion.java
package CasseraBottaBianchi;

public class EstrategaImplementacion implements Estratega {
    private static Ataque estrategiaAtaque = new Ataque();
    private static Defensa estrategiaDefensa = new Defensa();
    private static EstrategaImplementacion instance;

    private EstrategaImplementacion() {

    }

    public static EstrategaImplementacion getInstance() {
        // Patrón Singleton.
        if (instance == null) {
            instance = new EstrategaImplementacion();
        }
        return instance;
    }
    
    public Estrategia getEstrategia(CasseraBottaBianchiRobot robot) {
        if (robot.getEnergy() < 50 && robot.getOthers() < 3) {
            return estrategiaDefensa;
        }
        else if (robot.getEnergy() < 30) {
            return estrategiaDefensa;
        }
        else {
            return estrategiaAtaque;
        }
    }

    private static class Ataque implements Estrategia {
        public void run(CasseraBottaBianchiRobot robot) {
            robot.ahead(200); 
            robot.turnGunLeft(360);
            robot.back(10);
        }
        public void onScannedRobot(CasseraBottaBianchiRobot robot) {
            robot.fire(3);
        }
        public void onHitByBullet(CasseraBottaBianchiRobot robot) {
            robot.turnLeft(40);
        }
        public void onHitWall(CasseraBottaBianchiRobot robot) {
            robot.back(10);
            robot.turnRight(70);
            robot.ahead(400);
        }
    }
    private static class Defensa implements Estrategia {
        public void run(CasseraBottaBianchiRobot robot) {
            robot.back(100);
            robot.turnGunRight(120); 
            robot.back(140); 
            robot.turnGunLeft(120); 
        }
        public void onScannedRobot(CasseraBottaBianchiRobot robot) {
            robot.fire(1);
        }
        public void onHitByBullet(CasseraBottaBianchiRobot robot) {
            robot.back(20);
        }
        public void onHitWall(CasseraBottaBianchiRobot robot) {
            robot.ahead(20);
        }
    }
}

File: Talleres/4/src/EstrategaImplementacion.kt
package CasseraBottaBianchi

class EstrategaImplementacion private constructor() : Estratega {
    private val estrategiaAtaque = Ataque()
    private val estrategiaDefensa = Defensa()

    companion object {
        private var instance: EstrategaImplementacion? = null

        fun getInstance(): EstrategaImplementacion {
            if (instance == null) {
                instance = EstrategaImplementacion()
            }
            return instance!!
        }
    }

    override fun getEstrategia(robot: LaboRobot): Estrategia {
        return if (robot.getEnergy() < 50 && robot.getOthers() < 3) {
            estrategiaDefensa
        } else if (robot.getEnergy() < 30) {
            estrategiaDefensa
        } else {
            estrategiaAtaque
        }
    }

    inner class Ataque : Estrategia {
        override fun run(robot: LaboRobot) {
            robot.ahead(200)
            robot.turnGunLeft(360)
            robot.back(10)
        }
        override fun onScannedRobot(robot: LaboRobot) {
            robot.fire(3.0)
        }

        override fun onHitByBullet(robot: LaboRobot) {
            robot.turnLeft(40)
        }

        override fun onHitWall(robot: LaboRobot) {
            robot.back(10)
            robot.turnRight(70)
            robot.ahead(400)
        }
    }

    inner class Defensa : Estrategia {
        override fun run(robot: LaboRobot) {
            robot.back(100)
            robot.turnGunRight(120)
            robot.back(140)
            robot.turnGunLeft(120)
        }
        override fun onScannedRobot(robot: LaboRobot) {
            robot.fire(1.0)
        }

        override fun onHitByBullet(robot: LaboRobot) {
            robot.back(20)
        }

        override fun onHitWall(robot: LaboRobot) {
            robot.ahead(20)
        }
    }
}